### DISCUSSION TITLE
add plane-separate write_to_memory

### DISCUSSION THREAD

**Discussion by touste** [libvips/libvips#2061] [Ideas]
Some programming languages use the column-major format (ex: Julia, Fortran). It would be useful to have an option to write to memory in column-major format directly in vips in order to benefit from its fast processing capabilities.
A current workaround is to successively do a 90° rotation followed by a horizontal flip, but this only works for single-channel images.

Thanks!

**Comment by jcupitt** [comment]
Hello @touste,

It wouldn't be any faster than just calling rot/flip. I would just do:

```
data = image.rot90().fliphor().write_to_memory()
```

It should work for any image type. 

Do you need plane-separated images (eg. all the reds, then all the greens, then all the blues)? Yes, that could be made faster with a special operation.

**Comment by touste** [comment]
Thanks for the quick reply. 
Sorry I made a mistake in the title, what I want i column-major ordering (edited now). 

Yes this works well for single-channel images but unfortunately the memory layout is wrong for color images. Correct me if I'm wrong, but vips having a row-major storing order, the memory layout of a RGB image written to memory is the following:
[l1, c1, R] [l1, c1, G] [l1, c1, B] [l1, c2, R] [l1, c2, G] [l1, c2, B] ...

What I want is the following:
[l1, c1, R] [l2, c1, R] [l3, c1, R] ...[l1, c2, R] [l2, c2, R] [l3, c2, R] ... ... [l1, c1, G] [l2, c1, G] [l3, c1, G]

Which could be achieved by separating the channels into 3 images, rotating and flipping each image, and write the three images successively. I guess this is what you mean by plane-separated images?

**Comment by jcupitt** [comment]
That's right, row major and column major are the Fortran / C styles, and plane separate and plane interleaved are whether you have the minor (colour) dimension as the inner or outer one.

You could make a plane-separated image like this:

```python
y = [x.write_to_memory() for x in image.rot90().fliphor().bandsplit()]
```

Then pass that array of N memory objects to numpy.

You could also make a tall, thin image and then render that to memory:

```python
data = pyvips.Image.arrayjoin(image.rot90().fliphor().bandsplit(), across=1).write_to_memory()
```

Now in eg. numpy set the shape to [y, x, bands].

**Comment by touste** [comment]
Thanks! This works perfectly, and is still way faster than using the default loading scenario (in Julia, the default image loader uses ImageMagick).

**Comment by jcupitt** [comment]
I think a plane-separated `write_to_memory` could be useful. 

If you do bandsplit + arrayjoin, you'll end up evaluating the input N times as you render each plane out. Something like `write_to_memory_separated` could read the input once, then do a scattered write for each pixel, or perhaps each line of pixels.

Let's leave this open as an enhancement.
